### PR TITLE
runtime: throw exception in debug mode that condition is always true when fractionalPercent numerator > denominator

### LINKED ISSUE #11874: FractionalPercent when using with RouteMatch
**_Title:_** Log debug/info/whatever message when `FractionalPercent.numerator` greater than `FractionalPercent.denominator`
**_Description:_** When using `FractionalPercent` with `RouteMatch` it is possible to mess up with values on `denominator` / `numerator` so the underlying condition
```cpp
(random_value % denominator) < numerator
```
Will be always evaluated to `true` because of `numerator` exceeds `denominator`. It becomes hard to debug why configuration does not work (no logs, just shows that clusterN has been selected).

Do you guys think it is make sense to add some kind of logging to simplify configuration debuggability?
If so I can add this.

Thanks

### PR DESCRIPTION
**Pull request by belyalov**
Commit Message: Add debug log message when numerator > denominator
Additional Description: 
Risk Level: No Risk
Testing: No new tests added, was tested as part of troubleshooting of real issue
Docs Changes: Not Changed
Release Notes: Not added, maybe need to add?

Fixes #11874

### CODE REVIEW

File: source/common/runtime/runtime_impl.cc [hunk 1]
```diff
@@ -121,6 +121,19 @@ bool SnapshotImpl::featureEnabled(absl::string_view key,
...
+  // It becomes hard to debug why configuration does not work in case of wrong numerator.
+  // Log debug message that numerator is invalid.
+  uint64_t denominator_value =
+      ProtobufPercentHelper::fractionalPercentDenominatorToInt(percent.denominator());
+  if (percent.numerator() > denominator_value) {
+    ENVOY_LOG(debug,
+              "WARNING runtime key '{}': numerator ({}) > denominator ({}), condition always "
+              "evaluates to true",
+              key, percent.numerator(), denominator_value);
+    NOT_REACHED_GCOVR_EXCL_LINE;
```
Comment (belyalov) [inline source/common/runtime/runtime_impl.cc:134]: Is it ok have this as gcov exception or you'd prefer me to add super simple unittest?

File: source/common/runtime/runtime_impl.cc [hunk 2]
```diff
@@ -121,6 +121,19 @@ bool SnapshotImpl::featureEnabled(absl::string_view key,
...
+  }
```
Comment (tonya11en) [inline source/common/runtime/runtime_impl.cc:135]: I think this check should only occur in debug builds and just throw an exception. Something like:
```
#ifndef NDEBUG
if (percent.numerator() > denominator_value) {
   throw EnvoyException(.....);
}
#endif
```

File: source/common/runtime/runtime_impl.cc (hunk 1 above)
Comment (tonya11en) [reply source/common/runtime/runtime_impl.cc:134]: This change should have a unit test associated with it. Should be trivial.

Comment (belyalov) [reply source/common/runtime/runtime_impl.cc]: NP, added :)

**Comment by belyalov** [comment]
> Thanks! I left 2 comments on the patch. The changes should be easy.

Thanks for feedback!
Addresses all 2 comments.

File: source/common/runtime/runtime_impl.cc (hunk 2 above)
Comment (belyalov) [reply source/common/runtime/runtime_impl.cc]: Makes sense.

File: source/common/runtime/runtime_impl.cc [hunk 3]
```diff
@@ -121,6 +121,19 @@ bool SnapshotImpl::featureEnabled(absl::string_view key,
...
   }
 
+#ifndef NDEBUG
+  // When numerator > denominator condition is always evaluates to TRUE
+  // It becomes hard to debug why configuration does not work in case of wrong numerator.
+  // Log debug message that numerator is invalid.
+  uint64_t denominator_value =
+      ProtobufPercentHelper::fractionalPercentDenominatorToInt(percent.denominator());
+  if (percent.numerator() > denominator_value) {
+    throw EnvoyException(fmt::format("WARNING runtime key '{}': numerator ({}) > denominator ({}), "
```
Comment (tonya11en) [inline source/common/runtime/runtime_impl.cc:131]: I'd remove the WARNING part of this. You're halting the show with that exception, so it's not really a warning :)

Comment (belyalov) [reply source/common/runtime/runtime_impl.cc:131]: Oh yeah, sure, makes sense (I've just copy pasted it)

File: .bazelversion [hunk 4]
```diff
@@ -1 +0,0 @@
-3.3.1
```
Comment (tonya11en) [inline .bazelversion:1]: Should this be part of the diff?

Comment (belyalov) [reply .bazelversion:1]: oops, reverting it :)
Was testing it on several bazel versions

**Comment by belyalov** [comment]
@tonya11en ping? seems CI is now passing.. :)

File: test/common/runtime/runtime_impl_test.cc [hunk 5]
```diff
@@ -678,6 +678,29 @@ TEST_F(StaticLoaderImplTest, ProtoParsing) {
...
+    invalid_numerator:
+      numerator: 111
+      denominator: HUNDRED
+  )EOF");
+  setup();
+
+  envoy::type::v3::FractionalPercent fractional_percent;
+
+  // Numerator > denominator -> exception
+  EXPECT_CALL(generator_, random()).WillOnce(Return(500000));
```
Comment (tonya11en) [inline test/common/runtime/runtime_impl_test.cc:692]: Just pull this into the #ifndef and get rid of the #else block. That should be the last of it from my end.

**Comment by ggreenway** [comment]
I don't think having different error handling in debug vs release is a good idea. Can it just log instead, and do it regardless of compilation mode?

**Comment by belyalov** [comment]
> I don't think having different error handling in debug vs release is a good idea. Can it just log instead, and do it regardless of compilation mode?

This is what originally has been done.
@ggreenway @tonya11en let's finalize what's preferred way? :)

**Comment by ggreenway** [comment]
> > I don't think having different error handling in debug vs release is a good idea. Can it just log instead, and do it regardless of compilation mode?
> 
> This is what originally has been done.
> @ggreenway @tonya11en let's finalize what's preferred way? :)

Sorry about that; I didn't look at all the history when I reviewed it. I talked to @tonya11en and we agreed it makes most sense to log at debug level, in both debug and release builds. Again, sorry for the back-and-forth on this.

**Comment by belyalov** [comment]
@ggreenway ok, thanks for clarification.
Switched back to debug log message.

File: test/common/runtime/runtime_impl_test.cc [hunk 6]
```diff
@@ -678,6 +678,22 @@ TEST_F(StaticLoaderImplTest, ProtoParsing) {
...
+  )EOF");
+  setup();
+
+  envoy::type::v3::FractionalPercent fractional_percent;
+
+  // There is no assertion here - when numerator is invalid
+  // featureEnabled() will just drop debug log line.
+  EXPECT_CALL(generator_, random()).WillOnce(Return(500000));
+  loader_->snapshot().featureEnabled("invalid_numerator", fractional_percent);
+}
```
Comment (ggreenway) [inline test/common/runtime/runtime_impl_test.cc:695]: Please test for the log message. You'll probably need `Envoy::LogLevelSetter save_levels(spdlog::level::debug);` before the code that triggers the message, then use `EXPECT_LOG_CONTAINS("debug", "the text", loader_->snapshot().featureEnabled("invalid_numerator", fractional_percent));`

Comment (belyalov) [reply test/common/runtime/runtime_impl_test.cc:705]: Thanks! :)
I didn't know that you have such type of assertion, great!

**Comment by ggreenway** [comment]
Windows failure looks like an unrelated test flake.

**Comment by belyalov** [comment]
> Windows failure looks like an unrelated test flake.

:(
Do you want me to merge it with latest master or we can just re-start build?

**Comment by ggreenway** [comment]
> > Windows failure looks like an unrelated test flake.
> 
> :(
> Do you want me to merge it with latest master or we can just re-start build?

No need; merging as-is since the failure is unrelated.

**Pull request merged**